### PR TITLE
Added DAVE support

### PR DESCRIPTION
**Pull request by Wikijito7**
Added DAVE support to Kord lib

This PR is a little sloppy, but it works. I've tested it and it seems to accept DAVE correctly.

I'm open to modify anything, feel free to add comments with whatever needs to be modified further :).

This PR is based on [Discord4J's PR](https://github.com/Discord4J/Discord4J/pull/1344) where they implemented DAVE support too.

Solves #1013.

### CODE REVIEW

**Review by NoComment1105 — COMMENTED** [review]
Thanks for the PR. I've had an initial, cursory look at this (as I've not much spare time right now) and on the face of it, it seems pretty good! Only things I'll flag up is that there is already an existing `feature/dave` branch, could you rebase onto that please? And there seems to be some random full qualifications of clases, can these be removed? They seem unnnecessary (example: `dev.kord.voice.dave.NoOpDaveProtocol` in `VoiceConnection.kt`

You say you "managed to make it work"... does that indicate it's not fully tested yet and need stability testing or just a way of words?

Also to merge this, commits must be verified, could you look into that please? Organisation restriction :P

Once again thanks, I look forward to your response :D

**Comment by Wikijito7** [comment]
Hi! Thanks for the quick response!

I'll take a look at the build failing (seems weird, but I had some problems migrating the code from one computer onto this one, as the patched didn't want to _patch_ correctly). Also, the full qualification thingie is totally not intended, let me take a look at it tomorrow.

About the dave branch, ye, let me take a look at that branch and I'll try to rebase it. After I do that, would you like me to use that one as a base instead of main?

Also, I kinda expressed the "managed to make it work" bad. It works, although I found a bug after I created the PR, but I couldn't test it out as far as I would like, tbf, probably there might be some edge case or hidden bug that I need to refine, but the bot joins, plays and works with DAVE enabled.

Finally, about the verification thingie... _fair enough_. I'll take a look at it, I've changed computer recently and I didn't do it yet.

**Comment by psmarko** [comment]
Hey, thanks for the PR.

I gave it my first lookover and I understand this is a primarily a JVM implementation. That is expected but it has introduced more JVM-only usages than before. It would ease my mind a little if the voice module were converted into a multiplatform module, [like all the others](https://github.com/kordlib/kord/blob/e3ceb0f138e1d64cfcb74aa7e3d8f7ff576212be/rest/build.gradle.kts#L2), and all the JVM stuff (dependencies and code) is thrown into the JVM target. I bring this up since Kord is multiplatform oriented, and this PR, as is, takes voice a few steps back in that regard.

There are also several new usages of the Java stdlib for crypto and concurrency. This is not ideal.

I have not yet reviewed much logic nor other design.

Would you want to work on making this multiplatform friendly? If not, I can look into it myself. 
I appreciate the work here, cheers.

File: voice/build.gradle.kts [hunk 1]
```diff
@@ -18,4 +22,19 @@ dependencies {
...
+
+    // libdave-jvm for DAVE protocol — versions defined in libs.versions.toml
+    implementation(libs.libdave.api)
+    implementation(libs.libdave.impl.jni)
+
+    // Native libraries for libdave — include all platforms, loader picks the right one at runtime
+    runtimeOnly(libs.libdave.natives.darwin)
+    runtimeOnly(libs.libdave.natives.linuxX64)
+    runtimeOnly(libs.libdave.natives.linuxAarch64)
+    runtimeOnly(libs.libdave.natives.winX64)
```
Comment (psmarko) [inline voice/build.gradle.kts]: Kord should not bundle native artifacts. It is the user that's responsible for declaring platform-specific natives. These should rather be mentioned in the voice README.

```suggestion
```

**Review by psmarko — CHANGES_REQUESTED** [review]

File: voice/build.gradle.kts (hunk 1 above)
Comment (Wikijito7) [reply voice/build.gradle.kts]: Yea, fair. Gonna take a look and add it on the readme! Let me also see if this would break something of the current implementation.

**Comment by Wikijito7** [comment]
Hi psmarko! Thank you for the review!
> I gave it my first lookover and I understand this is a primarily a JVM implementation. That is expected but it has introduced more JVM-only usages than before. It would ease my mind a little if the voice module were converted into a multiplatform module, [like all the others](https://github.com/kordlib/kord/blob/e3ceb0f138e1d64cfcb74aa7e3d8f7ff576212be/rest/build.gradle.kts#L2), and all the JVM stuff (dependencies and code) is thrown into the JVM target. I bring this up since Kord is multiplatform oriented, and this PR, as is, takes voice a few steps back in that regard.

Yep, currently this supports JVM-only. I have to be honest, this is the only library that I found that could work with Kotlin's JVM capabilities, but I didn't take a deep look. If you know of any Kotlin-first library that wraps the C++ binary, I wouldn't mind changing into it. I tried making it myself but couldn't dedicate that much time with work-stuff.

> There are also several new usages of the Java stdlib for crypto and concurrency. This is not ideal.

Those usages are related with how the libdave-jvm work. I can change it, but I might need a little bit of guidance of which one you might like more.
 
> Would you want to work on making this multiplatform friendly? If not, I can look into it myself. I appreciate the work here, cheers.

I can try it, why not! Tbh I haven't worked on a Kotlin Native project yet, but I wouldn't mind trying out. Let me know if you have any preference regarding the previous points :).

**Comment by psmarko** [comment]
> this is the only library that I found that could work with Kotlin's JVM capabilities

It is one of the few, yes. My concern was there is little multiplatform scaffolding for future platform implementations. The roadmap for voice includes providing our own libdave mpp wrapper, so I wouldn't worry about the actual mpp implementation. Though I do want to make sure it is possible to add one in the future without breaking API (where possible).

> Those usages are related with how the libdave-jvm work.

While implementing the last encryption mode, I ran into AES256 not being exposed in the libsodium bindings we use for encryption. Thus I left it unimplemented. Is AES256 needed for DAVE? If so, I don't mind if the AES implementation stays for the time being, especially when not exposed publicly- just throw it into the JVM target.

`java.io.Closeable` can be replaced with a `close` function in `DaveProtocol`. This is how it is done in other places.

For concurrency usages, that's fine if it stays in the JVM target.

That's all the usages I spotted.

Just to reiterate, the goal is just to keep the voice common module platform-agnostic. A JVM implementation using libdave-jvm (`LibdaveDaveProtocol`) in the JVM target is fine.

Let me know if there's any issues turning it into a multiplatform module, no clue if there will be difficulties (though I would hope not).

Cheers.

**Comment by Wikijito7** [comment]
Hi again, 

I've done a partial migration of the module to a kmp-format, jvmMain, module. So far, it works. The thing is, I'm having a issue with voice on a particular flow, that may or may not happen but definitely is a bug. When the bot is kept alone on a VC playing (the other user leave) and rejoins, the bot voice stops working, as the epox changed. Moving the bot or reconnecting solves the error.  This is a new error introduced with how DAVE works.

I'll take a look at this error, try to fix it and then I'll update the PR with these changes, that way you can re-review it and tell me if this new reorder is closer to what you expect as a kmp-ready module.

  > The roadmap for voice includes providing our own libdave mpp wrapper, so I wouldn't worry about the actual mpp implementation.

 Got it, then we can work with this jvm-only variant for the mean time until your multiplatform-ready variant is finished, then we can migrate to that one.

**Comment by Wikijito7** [comment]
Sorry guys for the long-pause. Work's been crazy, today I remembered this was still undone.

I've uploaded some of the changes we talked, but the KMP migration was left undone. For the time being, I cannot dedicate the time I would like to finish everything perfectly.

Right now, this version is working fine. Probably part of the code can be done better, but it works as is.

Again, thank you for reviewing the previous version. Feel free to edit or comment whatever could be done better, and I'll try to take a look at it, although probably you'll be faster editing it by yourself if needed.

See you in the comments

File: voice/build.gradle.kts (hunk 1 above)
Comment (Wikijito7) [reply voice/build.gradle.kts]: Removed, tested on my local bot, adding these libraries as dependencies on the consumer-side keeps everything working.

**Comment by psmarko** [comment]
Just wanted to note that I've seen your update. Will find time to review in the near future, thanks.
